eth/downloader/stagedsync_stage_hashcheck.go

```diff
@@ -78,3 +103,259 @@ func (d *Downloader) unwindHashCheckStage(unwindPoint uint64) error {
 	}
 	return nil
 }
+
+func promoteHashedState(db ethdb.Database, progress uint64, datadir string) error {
+	if progress == 0 {
+		return promoteHashedStateCleanly(db, datadir)
+	}
+	return errors.New("incremental state promotion not implemented")
+}
+
+func promoteHashedStateCleanly(db ethdb.Database, datadir string) error {
+	err := copyBucket(datadir, db, dbutils.PlainStateBucket, dbutils.CurrentStateBucket, transformPlainStateKey)
+	if err != nil {
+		return err
+	}
+	return copyBucket(datadir, db, dbutils.PlainContractCodeBucket, dbutils.ContractCodeBucket, transformContractCodeKey)
+}
+
+func copyBucket(
+	datadir string,
+	db ethdb.Database,
+	sourceBucket,
+	destBucket []byte,
+	transformKeyFunc func([]byte) ([]byte, error)) error {
+
+	var m runtime.MemStats
+
+	buffer := newSortableBuffer()
+	files := []string{}
+
+	defer func() {
+		deleteFiles(files)
+	}()
+
+	err := db.Walk(sourceBucket, nil, 0, func(k, v []byte) (bool, error) {
+		newK, err := transformKeyFunc(k)
+		if err != nil {
+			return false, err
+		}
+		buffer.Put(newK, v)
+
+		bufferSize := buffer.Size()
+		if bufferSize >= buffer.OptimalSize {
+			sort.Sort(buffer)
+			file, err := buffer.FlushToDisk(datadir)
+			if err != nil {
+				return false, err
+			}
+			if len(file) > 0 {
+				files = append(files, file)
+			}
+
+			runtime.ReadMemStats(&m)
+
+			log.Info("Plain -> Hashed / created a buffer file",
+				"bucket", string(sourceBucket),
+				"name", file,
+				"size", bufferSize,
+				"currentKey", k,
```

Comment:
Please use something like `fmt.Sprintf("%x...", k[:4])` otherwise they are too long. We are only really interested to the first few nibbles in the log file